client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,814 @@
+	// The current active authorities.
+	currentAuthorities []Authority
+	// The current set id.
+	setId uint64
+	// Tree of pending standard changes across forks. Standard changes are
+	// enacted on finality and must be enacted (i.e. finalized) in-order across
+	// a given branch
+	pendingStandardChanges ForkTree[H, N]
+	// Pending forced changes across different forks (at most one per fork).
+	// Forced changes are enacted on block depth (not finality), for this
+	// reason only one forced hashNumber should exist per fork. When trying to
+	// apply forced changes we keep track of any pending standard changes that
+	// they may depend on, this is done by making sure that any pending hashNumber
+	// that is an ancestor of the forced changed and its effective block number
+	// is lower than the last finalized block (as signaled in the forced
+	// hashNumber) must be applied beforehand.
+	pendingForcedChanges []PendingChange[H, N]
+	// Track at which blocks the set id changed. This is useful when we need to prove finality for
+	// a given block since we can figure out what set the block belongs to and when the set
+	// started/ended.
+	authoritySetChanges AuthoritySetChanges[N]
+}
+
+// invalidAuthorityList authority sets must be non-empty and all weights must be greater than 0
+func invalidAuthorityList(authorities []Authority) bool {
+	if len(authorities) == 0 {
+		return true
+	}
+
+	for _, authority := range authorities {
+		if authority.Weight == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// NewGenesisAuthoritySet Get a genesis set with given authorities.
+func NewGenesisAuthoritySet[H comparable, N constraints.Unsigned](initial []Authority) (authSet *AuthoritySet[H, N]) {
+	if invalidAuthorityList(initial) {
+		return nil
+	}
+
+	return &AuthoritySet[H, N]{
+		currentAuthorities:     initial,
+		setId:                  0,
+		pendingStandardChanges: NewChangeTree[H, N](),
+		pendingForcedChanges:   make([]PendingChange[H, N], 0),
+		authoritySetChanges:    make(AuthoritySetChanges[N], 0),
+	}
+}
+
+// NewAuthoritySet creates a new AuthoritySet
+func NewAuthoritySet[H comparable, N constraints.Unsigned](authorities []Authority,
+	setId uint64,
+	pendingStandardChanges ForkTree[H, N],
+	pendingForcedChanges []PendingChange[H, N],
+	authoritySetChanges AuthoritySetChanges[N],
+) (authSet *AuthoritySet[H, N]) {
+	if invalidAuthorityList(authorities) {
+		return nil
+	}
+
+	return &AuthoritySet[H, N]{
+		currentAuthorities:     authorities,
+		setId:                  setId,
+		pendingStandardChanges: pendingStandardChanges,
+		pendingForcedChanges:   pendingForcedChanges,
+		authoritySetChanges:    authoritySetChanges,
+	}
+}
+
+// current Get the current set id and a reference to the current authority set.
+func (authSet *AuthoritySet[H, N]) current() (uint64, *[]Authority) {
+	return authSet.setId, &authSet.currentAuthorities
+}
+
+func (authSet *AuthoritySet[H, N]) revert() {
+	panic("AuthoritySet.revert not implemented yet")
```

Comment:
why wasn't this implemented?

Reply:
I had waited initially because it was never used in the tests, but I can implement and write my own tests

Reply:
This will be addressed in a different PR